### ISSUE TITLE
Specifying --sync-sgd with only one device aborts training

### ISSUE THREAD

**Issue by snukky** [marian-nmt/marian-dev#390]
For example, `--sync-sgd --devices 0`, or just `--sync-sgd` without specifying devices produces the following abort:

```
[2018-11-27 20:27:45] Error: Swap requires at least two graphs
[2018-11-27 20:27:45] Error: Aborted from virtual void marian::DefaultCommunicator::swapParams(const std::vector<std::shared_ptr<marian::TensorBase> >&) const in /path/to/marian-dev/src/training/communicator.h:206
```

I guess either it should fall back into singleton graph training or print a more useful message.

**Comment by frankseide** [comment]
We have 5 graph classes, and all have diverged and implement a slightly different subset of options. Therefore, I would not like this to fall back to singleton graph, but instead have `SyncGraphGroup` class to handle all (sync) cases.

Additionally, while we should definitely update `DefaultCommunicator`, we may also want to find out why this does not use the NCCL Communicator. The NCCL one should handle this.

**Comment by snukky** [comment]
It doesn't use NCCL Communicator because the library is not available on most machines I use and Marian is compiled without NCCL.

**Comment by emjotde** [comment]
@snukky did you fix this? I cannot reproduce it now.

**Comment by snukky** [comment]
Do you compile without NCCL?

**Comment by emjotde** [comment]
Yes, I just forgot that is happen during saving :) I am seeing it now.

**Comment by emjotde** [comment]
When I look at this, the only thing that's wrong, is the ABORT_IF. The code should just work.

**Comment by emjotde** [comment]
OK, the easiest way to fix this should be to implement a tensor-level element-wise swap. That will not require any additional memory allocator.

**Comment by emjotde** [comment]
@snukky do we have a regression test for this somewhere?

**Comment by snukky** [comment]
No, we don't. It's easy to add and I will add one to check the fix.

**Comment by emjotde** [comment]
Fixed in #397
